tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs

```diff
@@ -212,6 +212,48 @@ public void Dispose_should_raise_the_correct_events()
             _capturedEvents.Any().Should().BeFalse();
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void ReceiveMessage_should_throw_a_FormatException_when_message_is_an_invalid_size(
+            [Values(-1, 48000001)]
+            int length,
+            [Values(false, true)]
+            bool async)
+        {
+            using (var stream = new BlockingMemoryStream())
+            {
+                var bytes = BitConverter.GetBytes(length);
+                stream.Write(bytes, 0, bytes.Length);
+                stream.Seek(0, SeekOrigin.Begin);
+                var encoderSelector = new ReplyMessageEncoderSelector<BsonDocument>(BsonDocumentSerializer.Instance);
+
+                Exception exception;
+                if (async)
+                {
+                    _mockStreamFactory
+                        .Setup(f => f.CreateStreamAsync(_endPoint, CancellationToken.None))
+                        .ReturnsAsync(stream);
+                    _subject.OpenAsync(CancellationToken.None).GetAwaiter().GetResult();
+                    exception = Record
+                        .ExceptionAsync(async () => await _subject.ReceiveMessageAsync(10, encoderSelector, _messageEncoderSettings, CancellationToken.None))
```

Comment:
The `async` and `await` keywords seem unnecessary here.

Reply:
Doesn't `async ()` create an anonymous `async` function, which then enables the use of the `await` operator? So as written, both `async`/`await` should be needed.

Reply:
Oops, I think if we're going with this style for this ticket, then we'll need `ConfigureAwait(false)` as part of the `await`
```
                     exception = Record
                        .ExceptionAsync(async () => await _subject.ReceiveMessageAsync(10, encoderSelector, _messageEncoderSettings, CancellationToken.None).ConfigureAwait(false))
                        .GetAwaiter()
                        .GetResult();
```

Reply:
Let me rephrase in more detail:

`await` is not necessary because no further work is done after the async call

`async` is not necessary once we remove the `await`

Reply:
`ConfigureAwait(false)` is not necessary in the test projects.

It is of course mandatory in the non-test projects.

Reply:
Oh fair, we don't need it in the test projects (although I think it doesn't hurt?). Sorry @ksadoff, your code as fine as is!

Re: `async/await`: Oh very cool, you are indeed correct! It's required to have `async` if we want to use `await` but without `await`, of course we don't need `async`! 

(Side comment for gory in depth details in the interest of science/love of C# :) ):
I tried dropping the `await` call and got the following warnings, which of course we count as errors:
```
  BinaryConnectionTests.cs(238, 50): [CS1998] This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
  BinaryConnectionTests.cs(238, 54): [CS4014] Because this call is not awaited, execution of the current method continues before the call is completed. Consider applying the 'await' operator to the result of the call.
```

Running some extra tests for fun also confirmed that using `async/await` inside `Record.ExceptionAsync` and not using `async/await` inside `Record.ExceptionAsync` are equivalent :): 
```
 public void AsyncTest()
        {
            Task fail()
            {
                var tcs = new TaskCompletionSource<int>();
                tcs.SetException(new NotImplementedException());
                return tcs.Task;
            }
            var exception = Record
                .ExceptionAsync( async () =>  await fail())
                .GetAwaiter()
                .GetResult();
            exception.Should().BeOfType<NotImplementedException>();
            exception = Record
                .ExceptionAsync( () =>  fail())
                .GetAwaiter()
                .GetResult();
            exception.Should().BeOfType<NotImplementedException>();
        }
```